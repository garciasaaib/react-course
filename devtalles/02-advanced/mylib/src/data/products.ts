export interface Product {
	id: string;
	title: string;
	img: string;
}

export const product1 = {
	id: "1",
	title: "Coffee Mug - Card",
	img: "./coffee-mug.png",
};

export const product2 = {
	id: "2",
	title: "Coffee Mug - Meme",
	img: "./coffee-mug2.png",
};

export const products: Product[] = [product1, product2];
