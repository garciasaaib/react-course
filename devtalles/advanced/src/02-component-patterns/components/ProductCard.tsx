import styles from "../styles/styles.module.css";
import { useProduct } from "../hooks/useProduct";
import { createContext } from "react";
import { ProductContextProps, Props } from "../interfaces/interfaces";

// eslint-disable-next-line react-refresh/only-export-components
export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = ({ product, children }: Props) => {
	const { increaseBy, counter } = useProduct();
	return (
		<Provider value={{ counter, increaseBy, product }}>
			<div className={styles.productCard}>{children}</div>
		</Provider>
	);
};

