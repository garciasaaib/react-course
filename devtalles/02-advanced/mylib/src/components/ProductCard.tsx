import React, { CSSProperties, ReactElement } from "react";
import { useProduct } from "../hooks/useProduct";
import { Product } from "../data/products";
import { onChangeArgs } from "../hooks/useShoppingCart";
import { Provider } from "../context/ProductContext";
import { ProductCardHandlers } from "../interfaces/interfaces";
import styles from "../styles/styles.module.css";

// componente ProductCard
export interface InitialValues {
	count: number;
	maxCount: number;
}

export interface ProductCardProps {
	product: Product;
	children: (args: ProductCardHandlers) => ReactElement | ReactElement[];
	className?: string;
	style?: CSSProperties;
	onChange?: (args: onChangeArgs) => void;
	value?: number;
	initialValues?: InitialValues;
}

export const ProductCard = ({
	product,
	children,
	className,
	style,
	onChange,
	value,
	initialValues,
}: ProductCardProps) => {
	const { increaseBy, counter, maxCount, isMaxCountReached, reset } =
		useProduct({
			onChange,
			product,
			value,
			initialValues,
		});
	return (
		<Provider value={{ counter, increaseBy, product, maxCount }}>
			<div className={`${styles.productCard} ${className}`} style={style}>
				{children({
					product,
					maxCount,
					count: counter,
					isMaxCountReached,
					reset,
					increaseBy,
				})}
			</div>
		</Provider>
	);
};
