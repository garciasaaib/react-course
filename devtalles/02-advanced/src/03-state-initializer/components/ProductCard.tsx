import styles from "../styles/styles.module.css";
import { useProduct } from "../hooks/useProduct";
import { CSSProperties, ReactElement } from "react";
import { Product } from "../data/products";
import { onChangeArgs } from "../hooks/useShoppingCart";
import { Provider } from "../context/ProductContext";
import { ProductCardHandlers } from "../interfaces/interfaces";

// componente ProductCard
export interface InitialValues {
	count: number;
	maxCount: number;
}

export interface ProductCardProps {
	product: Product;
	children: (args: ProductCardHandlers) => ReactElement | ReactElement[];
	className?: string;
	style?: CSSProperties;
	onChange?: (args: onChangeArgs) => void;
	value?: number;
	initialValues?: InitialValues;
}

export const ProductCard = ({
	product,
	children,
	className,
	style,
	onChange,
	value,
	initialValues,
}: ProductCardProps) => {
	const { increaseBy, counter, maxCount, isMaxCountReached, reset } =
		useProduct({
			onChange,
			product,
			value,
			initialValues,
		});
	return (
		<Provider value={{ counter, increaseBy, product, maxCount }}>
			<div className={`${styles.productCard} ${className}`} style={style}>
				{children({
					product,
					maxCount,
					count: counter,
					isMaxCountReached,
					reset,
					increaseBy,
				})}
			</div>
		</Provider>
	);
};
