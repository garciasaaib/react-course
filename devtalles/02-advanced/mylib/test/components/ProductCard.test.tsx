import * as React from 'react';
import renderer from 'react-test-renderer';
import ProductCard from '../../src/components/index';
import { product1 } from '../../src/data/products';

describe('<ProductCard />', () => {
  it('should show the component with the product name', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>{() => <h1>ProductCard</h1>}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should show the component with the product name', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {({ count, increaseBy }) => (
          <>
            <h1>ProductCard</h1>
            <span>{count}</span>
            <button onClick={() => increaseBy(1)}>Increase</button>
          </>
        )}
      </ProductCard>
    );
    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();

    // tomamos el boton y simulamos un click
    (tree as any).children[2].props.onClick();

    tree = wrapper.toJSON();

    // probar que el span sea 1
    expect((tree as any).children[1].children[0]).toBe('1');
  });
});
