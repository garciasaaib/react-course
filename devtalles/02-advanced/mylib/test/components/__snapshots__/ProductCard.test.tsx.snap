// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard /> should show the component with the product name 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    ProductCard
  </h1>
</div>
`;

exports[`<ProductCard /> should show the component with the product name 2`] = `
<div
  className="productCard undefined"
>
  <h1>
    ProductCard
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    Increase
  </button>
</div>
`;
